File: benchmark/autolifter/single-pass/msp.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = -3;
Config SampleIntMax = 3;
Config NonLinear = true;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

msp =
  let mpp = fix (lambda f: Int -> List -> Int. \pre: Int. lambda l: List.
    match l with
      nil _ -> pre
    | cons {h, t} ->
      max pre (f (* h pre) t)
    end
  ) 1 in
  fix (\f: List -> Int. \xs: List.
  match xs with
    nil _ -> 1
  | cons {h, t} -> max (mpp xs) (f t)
  end
);

main = single_pass msp;

File: ml-benchmark/autolifter/single-pass/msp.ml
type list_ = Cons of int * list_ | Nil

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let msp =
  let rec f pre l =
    match l with
    | Nil -> pre
    | Cons (h, t) -> max pre (f (h * pre) t)
  in
  let mpp = f 1 in
  let rec g xs =
    match xs with
    | Nil -> 1
    | Cons (h, t) -> max (mpp xs) (g t)
  in
  g

let program = single_pass msp